### PR TITLE
Fixed definition of parsename, session_context and sp_set_session_context

### PR DESCRIPTION
**Pull request by rohit01010**
### Description
Following is the current definition of `PARSENAME`, `session_context` and `sp_set_session_context`
```
-- parsename
CREATE OR REPLACE FUNCTION sys.parsename(object_name sys.VARCHAR, object_piece int)
RETURNS sys.SYSNAME
AS 'babelfishpg_tsql', 'parsename'
LANGUAGE C IMMUTABLE STRICT; 

-- session_context
CREATE OR REPLACE FUNCTION sys.session_context ("@key" sys.sysname)
RETURNS sys.SQL_VARIANT 
AS 'babelfishpg_tsql', 'session_context' 
LANGUAGE C;
GRANT EXECUTE ON FUNCTION sys.session_context TO PUBLIC;

-- sp_set_session_context
CREATE OR REPLACE PROCEDURE sys.sp_set_session_context ("@key" sys.sysname, 
    "@value" sys.SQL_VARIANT, "@read_only" sys.bit = 0)
AS 'babelfishpg_tsql', 'sp_set_session_context'
LANGUAGE C;
GRANT EXECUTE ON PROCEDURE sys.sp_set_session_context TO PUBLIC;
```

But the correct definitions should be as follows, in these definitions `sys.NVARCHAR` should be used instead of `sys.VARCHAR` and `sys.NVARCHAR(128)` instead of `sys.SYSNAME`
```
--parsename
CREATE OR REPLACE FUNCTION sys.parsename(object_name sys.NVARCHAR(128), object_piece int)
RETURNS sys.NVARCHAR(128)
AS 'babelfishpg_tsql', 'parsename'
LANGUAGE C IMMUTABLE STRICT;

-- session_context
CREATE OR REPLACE FUNCTION sys.session_context ("@key" sys.NVARCHAR(128))
RETURNS sys.SQL_VARIANT 
AS 'babelfishpg_tsql', 'session_context'
LANGUAGE C;
GRANT EXECUTE ON FUNCTION sys.session_context TO PUBLIC;

-- sp_set_session_context
CREATE OR REPLACE PROCEDURE sys.sp_set_session_context ("@key" sys.NVARCHAR(128), 
    "@value" sys.SQL_VARIANT, "@read_only" sys.bit = 0)
AS 'babelfishpg_tsql', 'sp_set_session_context'
LANGUAGE C;
GRANT EXECUTE ON PROCEDURE sys.sp_set_session_context TO PUBLIC;
```

Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved

BABEL-4583

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** NA

* **Arbitrary inputs -** NA

* **Negative test cases -** NA

* **Minor version upgrade tests -** NA

* **Major version upgrade tests -** NA

* **Performance tests -** NA

* **Tooling impact -** NA

* **Client tests -** NA

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.4.0--3.5.0.sql [hunk 1]
```diff
@@ -59,6 +59,36 @@ WHERE t5.contype = 'p'
   AND ext.dbid = sys.db_id();
 
 
+ALTER FUNCTION sys.parsename(sys.VARCHAR, INT) RENAME TO parsename_deprecated_in_3_5_0;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.4.0--3.5.0.sql]: Lets use defensive coding here. Use try..catch block to achieve this

File: contrib/babelfishpg_common/src/babelfishpg_common.h [hunk 2]
```diff
@@ -70,5 +70,5 @@ typedef struct common_utility_plugin
 										  bool *isBaseDate, int *variantHeaderLen);
 	Oid			(*lookup_tsql_datatype_oid) (const char *typestr);
 	int32_t		(*GetUTF8CodePoint) (const unsigned char *in, int len, int *consumed_p);
-
```
Comment (Deepesh125) [inline contrib/babelfishpg_common/src/babelfishpg_common.h:73]: we need to make similar changes in pltsql.h also. Not sure why this changes are working in our testing

File: contrib/babelfishpg_tsql/src/session.c [hunk 3]
```diff
@@ -294,6 +295,12 @@ sp_set_session_context(PG_FUNCTION_ARGS)
 		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 						errmsg("The parameters supplied for the procedure \"sp_set_session_context\" are not valid.")));
 
+	encoded_key_bytelen = ((*common_utility_plugin_ptr->TsqlUTF8LengthInUTF16)(VARDATA_ANY(key_arg), VARSIZE_ANY_EXHDR(key_arg))) * 2;	/* Each UTF16 character is 2 bytes */
+
+	if (encoded_key_bytelen > 256)
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				errmsg("Cannot set key '%s' in the session context. The size of the key cannot exceed 256 bytes.", key)));
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/session.c:302]: As per T-SQL docs, maximum key size could be 128 bytes

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_common/src/babelfishpg_common.h (hunk 2 above)
Comment (rohit01010) [reply contrib/babelfishpg_common/src/babelfishpg_common.h:73]: Not required as definition of common_utility_plugin only exists in babelfishpg_common.h

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.4.0--3.5.0.sql (hunk 1 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.4.0--3.5.0.sql]: Added Exception statement to use defensive code here.

File: contrib/babelfishpg_tsql/src/session.c (hunk 3 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/session.c:302]: As discussed internally with you, we decided to go with 256.

**Review by Deepesh125 — APPROVED** [review]
